Documentation/contributing/testing/unit.rst

```diff
@@ -123,3 +108,120 @@ run this in a terminal next to your editor:
 
 This shell script depends on the ``inotify-tools`` package on Linux.
+
+.. _migrate-gocheck:
+
+Migrating tests off of ``gopkg.in/check.v1``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The Cilium codebase has historically used ``gopkg.in/check.v1`` to write unit and
+integration tests. Most of the functionality offered by that package is now
+present in `testing`_. What ``gocheck`` calls
+checkers are now available in `assert`_.
+
+There are still many tests written using ``gocheck`` today, which has some
+unfortunate consequences.
+
+1. Adding a test helper requires writing two versions: one for ``testing.T``
+   and one for ``check.C``.
+
+2. Executing those tests requires using non-standard flags.
+
+For this reason ``gopkg.in/check.v1`` has been switched to a fork called
+``github.com/cilium/checkmate``, which removes most of the problematic bits. Test
+executing is done using regular ``go test`` flags. More importantly, ``check.C``
+implements `testing.TB <https://pkg.go.dev/testing#TB>`__.
+
+To convert a test from ``gocheck`` you can use the following approach:
+
+1. Replace ``SetUp`` fixtures with helpers that take ``testing.TB``.
+
+2. Replace ``TearDown`` fixtures with calls to ``testing.TB.Cleanup()``,
+   possibly in the helper you added to replace ``SetUp``.
+
+3. Replace calls to ``c.Assert`` with `assert`_.
+
+4. Replace tests methods with a normal ``func TestHelloWorld(t *testing.T)``
+   function.
+
+Let's take an example from the ``gocheck`` documentation and migrate it:
+
+.. code-block:: go
+
+    package hello_test
+
+    import (
+        "testing"
+
+        . "gopkg.in/check.v1"
+    )
+
+    func Test(t *testing.T) { TestingT(t) }
+
+    type MySuite struct{}
+
+    var _ = Suite(&MySuite{})
+
+    func (s *MySuite) SetUpTest(c *C) {
+        // setup code
+    }
+
+    func (s *MySuite) TearDownTest(c *C) {
+        // cleanup code
+    }
+
+    func (s *MySuite) TestHelloWorld(c *C) {
+        c.Assert(42, Equals, "42")
+    }
+
+    type SomeOtherSuite struct{}
+
+    var _ = Suite(&SomeOtherSuite{})
+
+    func (s *SomeOtherSuite) TestTheRealAnswer(c *C) {
+        c.Assert("42", Equals, "42")
+    }
+
+After applying the above rules we might end up with something like this:
+
+.. code-block:: go
+
+    package hello_test
+
+    import (
+        "testing"
+
+        . "github.com/cilium/checkmate"
+        "github.com/stretchr/testify/assert"
+    )
+
+    func Test(t *testing.T) { TestingT(t) }
+
+    func setupHelper(tb testing.TB) {
+        // setup code
+
+        tb.Cleanup(func() {
+            // cleanup code
+        })
+    }
+
+    func TestHelloWorld(t *testing.T) {
+        setupHelper(t)
+
+        assert.Equal(t, 42, "42")
+    }
+
+    type SomeOtherSuite struct{}
+
+    var _ = Suite(&SomeOtherSuite{})
+
+    func (s *SomeOtherSuite) TestTheRealAnswer(c *C) {
+        assert.Equal(c, "42", "42")
+    }
+
+As you can see we didn't get round to converting all tests yet, but that is OK.
```

Comment:
I really wish we'd move away from the 'suite' concept altogether. It would be useful to reaffirm this in an additional short paragraph. Currently, this could be interpreted as "adding tests leaning on checkmate is OK". If you're working on extending test coverage, taking some extra time to lift a suite out of check.C would be much appreciated.

Reply:
Agreed and this isn't meant to encourage using suite. Reworded a bit and shuffled some of the doc around.